projects/media-viewer/src/lib/media-viewer/viewers/image-viewer/image-viewer.component.ts

```diff
@@ -1,52 +1,60 @@
-import { Component, Input, Renderer2, ViewChild, ElementRef, SimpleChanges, OnChanges } from '@angular/core';
+import { Component, Input, Renderer2, ViewChild, ElementRef } from '@angular/core';
 import { EmLoggerService } from '../../../logging/em-logger.service';
-import { GenericOperation, RotateOperation, ZoomOperation } from '../../media-viewer.model';
+import {DownloadOperation, PrintOperation, RotateOperation, ZoomOperation} from '../../media-viewer.model';
 
 @Component({
     selector: 'app-image-viewer',
     templateUrl: './image-viewer.component.html',
     styleUrls: ['./image-viewer.component.scss'],
 })
-export class ImageViewerComponent implements OnChanges {
+export class ImageViewerComponent {
 
   @Input() url: string;
   @Input() downloadFileName: string;
   @Input() originalUrl: string;
   @ViewChild('img') img: ElementRef;
   rotation = 0;
 
-  @Input() rotateOperation: RotateOperation;
-  @Input() zoomOperation: ZoomOperation;
-  @Input() printOperation: GenericOperation;
-
   constructor(private renderer: Renderer2,
               private log: EmLoggerService) {
   }
 
-  ngOnChanges(changes: SimpleChanges): void {
-    for (let change in changes) {
-      let operation = changes[change].currentValue;
-      if(operation && operation.action) {
-        this[operation.action].call(this, operation);
-      }
-    }
-  }
 
-  rotate(operation: RotateOperation) {
+  @Input()
+  set rotateOperation(operation: RotateOperation) {
+    if (operation) {
```

Comment:
this is a bit nonsensical - if the type declaration of operation is RotateOperation then it will always exist. This was not true in typescript 2.x (and you can disable it in 3.x) but types are non-nullable by default. If you want a nullable type the declaration needs to be `RotateOperation | null`

Reply:
if you want an optional type then you add `?` which is the same as `| undefined`

Reply:
the first time these functions are called, when the component is first loaded, the value seems to be null

Reply:
set rotateOperation is a way to define a setter - I can't make the parameter optional. I added the if statement because it's executed with the argument == null - I'm guessing when it's initialised.

Reply:
so without the if statement the print would execute when the component was being initialised. Any ideas how to make it prettier ?

Reply:
personally, I would look into why a null is being emitted... that is super weird

Reply:
if you absolutely can't work it out, then change the type definition to reflect that the value might be null

Reply:
it's not that a null is being emitted. it's the async pipe evaluating to null when the subject value hasn't resolved yet

Reply:
Why would it emit something before it's being resolved?

Reply:
the subject isn't emitting null. it's the angular async pipe that returns null when no value has been received from the subject yet.
here's a proposal/bug that's been raised to try and get them to change this behaviour
https://github.com/angular/angular/issues/16982

Reply:
leaky abstraction much? okay, let's update the type info so it makes sense then merge it

Reply:
would it be cleaner to have non-null pipe as a workaround? 

[zoomOperation]="zoom | async | non-null"

Reply:
if that's possible it would be good - I've  just pushed a change to update the type info